File: NeetCode/Two Pointers/threeSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class threeSum {
    public static List<List<Integer>> calculateThreeSum(int[] nums) {
        int target = 0;
        Arrays.sort(nums);                                                      //! Time Complexity = O(nlog n)

        Set<List<Integer>> s = new HashSet<>(); 
        List<List<Integer>> output = new ArrayList<>();

        for(int i = 0;i < nums.length;i++) {                                    //! Time Complexity = O(n)
            int j = i+1;
            int k = nums.length-1;

            while(j < k) {                                                      //! Time Complexity = O(n)
                int sum = nums[i] + nums[j] + nums[k];
                if(sum == target) {
                    s.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;
                } else if(sum < target) {
                    j++;
                } else {
                    k--;
                }
            }
        }
        output.addAll(s);
        
        return output;
    }                                                                           //! Total Time Complexity = O(n^2)

    public static void main(String[] args) {
        int[] nums = {-1,0,1,2,-1,-4};
        
        System.out.println(calculateThreeSum(nums));
    }
}

File: NeetCode/Two Pointers/ThreeSum.cs
public class Solution {
    public List<List<int>> ThreeSum(int[] nums) {
        Array.Sort(nums);

        HashSet<Tuple<int, int, int>> uniqueTriplets = new HashSet<Tuple<int, int, int>>();
        List<List<int>> res = new List<List<int>>();
        
        for (int i = 0;i < nums.Length;i++)
        {
            int j = i+1;
            int k = nums.Length-1;
            int target = -nums[i];
            while(j < k)
            {
                if (nums[j] + nums[k] == target)
                {
                    var tuple = Tuple.Create(nums[i], nums[j], nums[k]);
                    uniqueTriplets.Add(tuple);
                    j++;
                }
                else if (nums[j] + nums[k] > target)
                {
                    k--;
                }
                else 
                {
                    j++;
                }
            }
        }

        foreach (var triplet in uniqueTriplets)
        {
            res.Add(new List<int> { triplet.Item1, triplet.Item2, triplet.Item3 });
        }

        return res;
    }
}
